src/IRM/IRMAdaptiveLinearKink.sol

```diff
@@ -0,0 +1,173 @@
+    struct IRState {
+        /// @dev The current rate at kink.
+        int208 kinkRate;
+        /// @dev The timestamp of the last update to the model.
+        uint48 lastUpdate;
+    }
+
+    /// @notice Get the cached state of a vault's irm.
+    /// @return kinkRate The last computed rate at kink.
+    /// @return lastUpdate The last update timestamp.
+    /// @dev Note that this state may be outdated. Use `computeInterestRateView` for the latest interest rate.
+    mapping(address => IRState) public irState;
+
+    /// @notice Deploy IRMAdaptiveLinearKink
+    /// @param _kink The utilization rate targeted by the interest rate model.
+    /// @param _initialKinkRate The initial interest rate at kink.
+    /// @param _minKinkRate The minimum interest rate at kink that the model can adjust to.
+    /// @param _maxKinkRate The maximum interest rate at kink that the model can adjust to.
+    /// @param _slope The steepness of interest rate function below and above the kink.
+    /// @param _adjustmentSpeed The speed at which the rate at kink is adjusted up or down.
+    constructor(
+        int256 _kink,
+        int256 _initialKinkRate,
+        int256 _minKinkRate,
+        int256 _maxKinkRate,
+        int256 _slope,
+        int256 _adjustmentSpeed
+    ) {
+        kink = _kink;
+        initialKinkRate = _initialKinkRate;
+        minKinkRate = _minKinkRate;
+        maxKinkRate = _maxKinkRate;
+        slope = _slope;
+        adjustmentSpeed = _adjustmentSpeed;
+    }
+
+    /// @inheritdoc IIRM
+    function computeInterestRate(address vault, uint256 cash, uint256 borrows) external returns (uint256) {
+        if (msg.sender != vault) revert E_IRMUpdateUnauthorized();
+        (uint256 avgRate, int256 endKinkRate) = computeInterestRateInternal(vault, cash, borrows);
+        irState[vault] = IRState(int208(endKinkRate), uint48(block.timestamp));
+        return avgRate;
+    }
+
+    /// @inheritdoc IIRM
+    function computeInterestRateView(address vault, uint256 cash, uint256 borrows) external view returns (uint256) {
+        (uint256 avgRate,) = computeInterestRateInternal(vault, cash, borrows);
+        return avgRate;
+    }
+
+    /// @notice Compute the new interest rate and rate at kink of a vault.
+    /// @param vault Address of the vault to compute the new interest rate for.
+    /// @param cash Amount of assets held directly by the vault.
+    /// @param borrows Amount of assets lent out to borrowers by the vault.
+    /// @return The new interest rate at current utilization.
+    /// @return The new interest rate at kink.
+    function computeInterestRateInternal(address vault, uint256 cash, uint256 borrows)
+        internal
+        view
+        returns (uint256, int256)
+    {
+        IRState memory state = irState[vault];
+        // Initialize rate if this is the first interaction.
+        if (state.lastUpdate == 0) return (uint256(initialKinkRate), initialKinkRate);
```

Comment:
Shouldn't the first return value should be  the IR at the utilization, given initialKinkRate: `calcLinearKinkRate(initialKinkRate, err)`?

Reply:
yes, good catch